llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+	explicit SubGraph(const llvm::BasicBlock* Entry, BlockSet Blocks):
+		Entry(Entry), Blocks(std::move(Blocks))
+	{
+	}
+	const BasicBlock* getEntry() const
+	{
+		return Entry;
+	}
+private:
+	GraphNode* getOrCreate(const BasicBlock* BB)
+	{
+		auto it = Nodes.find(BB);
+		if (it == Nodes.end())
+		{
+			it = Nodes.emplace(BB, GraphNode(BB, *this)).first;
+		}
+		return &it->second;
+	}
+	friend struct GraphTraits<SubGraph*>;
+	friend struct GraphNode;
+
+	const BasicBlock* Entry;
+	BlockSet Blocks;
+	NodeMap Nodes;
+};
+
+GraphNode::GraphNode(const BasicBlock* BB, SubGraph& Graph): BB(BB), Graph(Graph)
+{
+        for (auto Succ: successors(BB))
+        {
+                // Skip edges that go outside of the SubGraph
+                if (!Graph.Blocks.count(Succ))
+                        continue;
+		if (Succ == Graph.getEntry())
+			continue;
+                Succs.push_back(Succ);
+        }
+}
+
+template <> struct GraphTraits<SubGraph*> {
+        typedef GraphNode NodeType;
+        typedef NodeType* NodeRef;
+        typedef mapped_iterator<SmallVectorImpl<const BasicBlock*>::iterator, std::function<GraphNode*(const BasicBlock*)>> ChildIteratorType;
+
+        static NodeType *getEntryNode(SubGraph* G) { return G->getOrCreate(G->Entry); }
+        static inline ChildIteratorType child_begin(NodeType *N) {
+                return ChildIteratorType(N->Succs.begin(), [N](const BasicBlock* BB){ return N->Graph.getOrCreate(BB);});
+        }
+        static inline ChildIteratorType child_end(NodeType *N) {
+                return ChildIteratorType(N->Succs.end(), [](const BasicBlock* BB){ llvm_unreachable("dereferencing past-the-end iterator");return nullptr;});
+        }
+};
+}//namespace llvm
+
+namespace cheerp
+{
+
+class BasicBlockGroupData
+{
+	FunctionData& data;
+	DeterministicBBSet inner;
+	bool multiHead {false};
+	const llvm::BasicBlock* start;
+	const llvm::BasicBlock* from;
+	//TODO(carlo): an optmization might be having from be a set<BasicBlock>, conserving the phi that are equals
+	BasicBlockGroupData* parent;
+	std::deque<BasicBlockGroupData> subGroups;
+	std::map<const llvm::BasicBlock*, int> subGroupsID;
+	bool visitingAll{false};
+	int iterationCount{-1};
+	std::vector<std::pair<const llvm::BasicBlock*, int> > escaping;
+	static const DeterministicBBSet getAllBasicBlocks(const llvm::Function& F)
+	{
+		DeterministicBBSet set;
+		for (const llvm::BasicBlock& bb : F)
+		{
+			set.insert(&bb);
+		}
+		return set;
+	}
+	void splitIntoSCCs(std::deque<BasicBlockGroupData>& blockQueue, std::map<const llvm::BasicBlock*, int>& blockToIndexMap);
+public:
+	BasicBlockGroupData(FunctionData& data, BasicBlockGroupData* parentBBGD, const DeterministicBBSet& inner, const llvm::BasicBlock* start = nullptr)
+		: data(data), inner(inner), start(start), from(nullptr), parent(parentBBGD)
+	{
+		if (start)
+			assert(start->getParent() == data.getFunction());
+	}
+	BasicBlockGroupData(FunctionData& data)
```

Comment:
Are these constructors actually used?

Reply:
Yes, I checked

Reply:
Adding a comment on the with tricky behaviour